llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp

```diff
@@ -30,6 +31,7 @@
 #include "llvm/Transforms/Utils/BasicBlockUtils.h"
 #include "llvm/Transforms/Utils/LoopUtils.h"
 #include "llvm/Transforms/Utils/ScalarEvolutionExpander.h"
+#include "llvm/Transforms/Vectorize/LoopVectorizationLegality.h"
```

Comment:
Might be better to define HistogramInfo in VPlan.h instead of pulling `LoopVectorizationLegality.h` in here, as VPlanRecipes.cpp should not depend/rely on LVL

Reply:
Sadly, that's not quite possible at the moment. LoopVectorizationLegality.h needs the definition of HistogramInfo since it creates a SmallVector of them, and accesses the fields. I can't include VPlan.h in the other header, since VPlan.h is under the lib/ subtree instead of the include/ subtree. I could perhaps try to fudge it by always including VPlan.h first in any .cpp file, but that feels fragile (and clang-format doesn't like it anyway).

I suspect a shared header under include/ would be best. Or have I missed something?

Reply:
It looks like VPHistogramRecipe only needs `Update` of `HistogramInfo` and that only for retrieving the legacy cost. Ideally we would include `VPHistogramRecipe::computeCost` in this patch, avoiding the need for accessing `Update`. 

If that isn't easily possibly, can we just store `Update` (with a comment what it is used for)?

Reply:
Still pending?

Reply:
This was done with the last round of updates. Old diff?